### PR TITLE
Prevent ActiveStorage blob from purging if it has shared attachments.

### PR DESCRIPTION
**Pull request by jaspervanbrian**
### **Summary**
Issue #33013 states if you purge an attachment, the blob is destroyed
instantaneously without checking if it has existing attachments.
This behaviour makes a person write more comparison code to determine
if it will use `purge` or `detach` which is unnecessary and verbose.

So, to prevent this, I slightly modified the `purge` method on attachment.rb:

    def purge
      delete
      blob.purge if blob.attachments.blank?
    end

Same for the `purge_later` method, but the condition check is happening on
the PurgeJob job.

    def perform(blob)
      blob.purge if blob.attachments.blank?
    end

### CODE REVIEW

**Comment by georgeclaghorn** [comment]
This implementation is susceptible to a race condition that can lead to data loss: process A may see a blob as having no attachments and proceed to purge it, while concurrent process B may attach the same blob to a record after process A’s attachments check and before its purge call.

#33405 added a foreign-key constraint to the attachments table for blobs. I think we can just rescue `ActiveRecord::ForeignKey` in `ActiveStorage::Blob#purge` and do nothing. `ActiveStorage::PurgeJob` already does that—I considered rescuing in `#purge` and decided against it, but I’m coming around to it now.

**Comment by jaspervanbrian** [comment]
So... should I include `rescue` in both `#purge` and `ActiveStorage::PurgeJob#perform`?

**Comment by jaspervanbrian** [comment]
So, I decided to only amend the changes. It seems the problem here lies with how blob handles `ActiveRecord::InvalidForeignKey` violation.

**Comment by georgeclaghorn** [comment]
There’s an outdated test that needs updating:

```
Failure:
ActiveStorage::BlobTest#test_purge_fails_when_attachments_exist [/home/travis/build/rails/rails/activestorage/test/models/blob_test.rb:180]:
ActiveRecord::InvalidForeignKey expected but nothing was raised.
```

**Comment by jaspervanbrian** [comment]
So I just amended the changes I made and changed the commit message to make things clearer. I will update the outdated test right now.

File: activestorage/test/models/blob_test.rb [hunk 1]
```diff
@@ -177,7 +177,7 @@ class ActiveStorage::BlobTest < ActiveSupport::TestCase
   test "purge fails when attachments exist" do
     create_blob.tap do |blob|
       User.create! name: "DHH", avatar: blob
-      assert_raises(ActiveRecord::InvalidForeignKey) { blob.purge }
+      assert_nothing_raised { blob.purge }
```
Comment (georgeclaghorn) [inline activestorage/test/models/blob_test.rb]: Let’s assert that the blob isn’t deleted instead:

```ruby
assert_no_difference(-> { ActiveStorage::Blob.count }) { blob.purge }
```

Finally, please update the name of the test to reflect that `purge` no longer fails when attachments exist: `"purge does nothing when attachments exist"`.

Comment (jaspervanbrian) [reply activestorage/test/models/blob_test.rb]: Oh okay, that seems to be more appropriate! 👍

**Comment by padi** [comment]
Nice work on your first Rails contribution @jaspervanbrian!

**Pull request merged**